zmq/cl-zmq.py

```diff
@@ -0,0 +1,207 @@
+            endpoint_opt = nt.endpoint_option()
+            endpoint = options[endpoint_opt]
+            if endpoint != "null":
+                if len([1 for prefix in ZMQ_TRANSPORT_PREFIXES if
+                        endpoint.startswith(prefix)]) != 0:
+                    continue
+                yield endpoint
+
+    def check_option_warnings(options, plugin):
+        if not Setup._at_least_one_binding(options):
+            plugin.log("No zmq publish sockets are bound as per launch args",
+                       level='warn')
+        for endpoint in Setup._iter_endpoints_not_ok(options):
+            plugin.log(("Endpoint option {} doesn't appear to be recognized"
+                       ).format(endpoint), level='warn')
+
+    ###########################################################################
+
+    def _iter_endpoint_setup(options):
+        for nt in NOTIFICATION_TYPES:
+            endpoint_opt = nt.endpoint_option()
+            if options[endpoint_opt] == "null":
+                continue
+            endpoint = options[endpoint_opt]
+            hwm_opt = nt.hwm_option()
+            hwm = int(options[hwm_opt])
+            yield endpoint, nt, hwm
+
+    def get_setup_dict(options):
+        setup = {}
+        for e, nt, hwm in Setup._iter_endpoint_setup(options):
+            if e not in setup:
+                setup[e] = {'notification_type_names': [],
+                            'high_water_mark':         hwm}
+            setup[e]['notification_type_names'].append(str(nt))
+            # use the lowest high water mark given for the endpoint
+            setup[e]['high_water_mark'] = min(
+                setup[e]['high_water_mark'], hwm)
+        return setup
+
+    ###########################################################################
+
+    def log_setup_dict(setup, plugin):
+        for e, s in setup.items():
+            m = ("Endpoint {} will get events from {} subscriptions "
+                 "published with high water mark {}")
+            m = m.format(e, s['notification_type_names'], s['high_water_mark'])
+            plugin.log(m)
+
+
+###############################################################################
+
+plugin = Plugin()
+
+@plugin.init()
+def init(options, configuration, plugin, **kwargs):
+    Setup.check_option_warnings(options, plugin)
+    setup_dict = Setup.get_setup_dict(options)
+    Setup.log_setup_dict(setup_dict, plugin)
+    reactor.callFromThread(publisher.load_setup, setup_dict)
+
+SUBSCRIBE_TEMPLATE = """
+@plugin.subscribe("{}")
+def on_{}(plugin, *args, **kwargs):
+    if len(args) != 0:
+        plugin.log("got unexpected args: {}".format(args), level="warn")
+    reactor.callFromThread(publisher.publish_notification, "{}",
+                           *args, **kwargs)
+"""
+
+DEFAULT_HIGH_WATER_MARK = 1000
+
+for nt in NOTIFICATION_TYPES:
+    # subscribe to all notifications
+    subscribe = SUBSCRIBE_TEMPLATE.format(nt, nt, nt, nt)
+    exec(subscribe)
```

Comment:
This seems rather dangerous. I'd suggest creating a generic handler, use `functools.partial` to generate variants with the notification type bound and then using `plugin.add_subscription` to register them.

Something like the following would do I think:

```python
def on_notification(notification_type, plugin, *args, **kwargs):
    if len(args) != 0:
￼        plugin.log("got unexpected args: {}".format(args), level="warn")
￼    reactor.callFromThread(publisher.publish_notification, "{}",
￼                           *args, **kwargs)

for nt in NOTIFICATION_TYPES:
    f = functools.partial(on_notification, nt)
    plugin.add_subscription(nt, f)
```

This would not call the python interpreter with a variable payload. In this case it's likely ok with your mechanism, since it just fills in the notification types, but in other cases this could be a security vulnerability since you execute code that was provided from outside.